### ISSUE TITLE
Investigate and probably use Apache Lucy for searching

### ISSUE THREAD

**Issue by pombase-admin** [pombase/canto#584]
We currently use "CLucene" - a C++ port of Lucene (which is Java) to do indexing, searching and completion in Canto.  CLucene has a Perl wrapper which makes life easy.

Unfortunately CLucene and the Perl wrapper library are now unmaintained.  They aren't getting updates and they don't work out of the box on the latest version of Ubuntu Linux (and probably others).  This is likely to cause us and others problems when installing Canto or updating the operating system of servers running Canto.  I noticed the problem when I updated to the latest Ubuntu on my laptop.  It took a bit of fiddling around to get it working.  That's likely to be a problem on oliver0 too at some point.

There is now a similar project called "Lucy" with is a C library with Perl wrapper.  It seems to be very well maintained and bugs get fixed quickly (from a quick Google search).

Lucy is likely to work in a similar but not identical way to CLucene.  There's bound to be some problems when we start using it.  I think it needs doing long term though.  Since it's still being improved, lt's likely to result in better searches once we get through the pain.

**Comment by ValWood** [comment]
In an older ticket Antonia mentioned 'fuzzy' searching. This might be covered here, but I don't think its a big issue.
Synonyms is the way to go.

**Comment by ValWood** [comment]
Is this still on the table?
I noticed there are other tickets about using Lucene?

**Comment by kimrutherford** [comment]
> Is this still on the table?

I think so.  The current package we use (CLucene) is a maintenance burden.  Given that we plan to use the main PomBase site for autocompletion at some point (#1338) I'd like swap CLucene for "Lucy".  It looks like Lucy isn't quite as clever as CLucene but it's in development and has a maintainer.

It would be good to have something basic but usable (Lucy) for PHI-base and others who use Canto.  We can implement any fancy stuff we need in the main PomBase site (it already has fuzzy searching for example).

CLucene and Lucy are clones of Lucene but they can't do everything the real Lucene can.  The main PomBase site uses the real Lucene (via Solr).

I'll try to get to this soon-ish.

**Comment by kimrutherford** [comment]
I've made some progress on this but there are still some major things not working.  I'm making the changes on a branch so I don't muck up the main Canto.

The main thing that still needs fixing is querying sub-ontologies.  I should be able to get that fixed in early January.

@henryosim once this is fixed we'll be able to run Canto without using Docker which will be much nicer

**Comment by kimrutherford** [comment]
> Lucy isn't quite as clever as CLucene

I did quite a bit of work on this to get around the parts of CLucene that are missing from Lucy.  Unfortunately I've had to give up.  There are too too many holes in the functionality.

The Perl bindings for current library (CLucene) work, just, on Debian and Ubuntu systems but not on other Linux distributions.  Debian is what we use on oliver1 and I use Ubuntu on my desktop so PomBase Canto is OK for now.

I think we should encourage anyone who wants to use Canto to install Debian or Ubuntu on their server.

So I'm going to close this.

Longer term I still think we should use the main PomBase site for autocompletion.  The main site uses real Apache Lucene/Solr and should handle everything we need.  That's a separate issue (#1338).
